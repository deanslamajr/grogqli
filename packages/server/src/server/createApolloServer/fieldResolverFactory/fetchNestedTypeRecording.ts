import {
  TypeNameToIdMapping,
  getTypeIdFromTypeName,
  getTypeRecording,
} from '../../files/type';

interface FetchNestedTypeRecordingParams {
  recordingId: string;
  typeName: string;
  typeNameToIdMappingData: TypeNameToIdMapping;
}

// TODO handle args
// TODO optimization: should only need to do this once for all root level fields of the given query execution
//  * wait for semaphore access (only allows a single access at any given time) https://www.npmjs.com/package/await-semaphore
//  * after acquiring semaphore, check cache for resolved value
//  * If exists, release semaphore and return value
//  * Else, do the work below, set cache, release semaphore, and return value
export const fetchNestedTypeRecording = async ({
  recordingId,
  typeName,
  typeNameToIdMappingData,
}: FetchNestedTypeRecordingParams) => {
  const typeId = await getTypeIdFromTypeName({
    typeName,
    typeNameToIdMappingData,
  });
  // TODO handle the case where the given typeName does not exist in the given schemaId's types.json
  if (typeId === null) {
    throw new Error(`
      TODO handle the case where the given typeName does not exist in the given types.json.
      typeName:${typeName}
    `);
  }

  const typeRecording = await getTypeRecording({
    args: {}, // TODO: replace hardcoded args value (this was done to make compiler pass)
    typeId,
    recordingId,
  });
  return typeRecording; //.value;
};
