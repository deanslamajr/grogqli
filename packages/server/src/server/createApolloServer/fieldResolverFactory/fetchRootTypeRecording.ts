import {
  getTypeIdFromTypeName,
  getTypeRecording,
  TypeNameToIdMapping,
} from '../../files/type';
import {
  getWorkflowById,
  WorkflowRecordingsFile,
  WorkflowRecordingsFileVersion1,
} from '../../files/workflow';
import {
  getOperationFile,
  OperationRecordingsFileVersion1,
  OperationNameToIdMapping,
} from '../../files/operation';

interface GetOpDataParams {
  operationsData: OperationNameToIdMapping;
  opName: string;
}

interface GetRootTypeRecordingIdParams {
  opId: string;
  workflowId: string;
}

const getOpIdFromOpName = ({
  operationsData,
  opName,
}: GetOpDataParams): string => {
  const opData = operationsData.operations[opName];

  // TODO handle case where a opId does not exist for the given opName
  if (opData === undefined) {
    throw new Error(
      `TODO handle case where an opId does not exist for the given opName. opName:${opName}`
    );
  }

  return opData.id;
};

const getOpRecordingIdFromWorkflow = async ({
  opId,
  workflowId,
}: GetRootTypeRecordingIdParams): Promise<string> => {
  const workflowData: WorkflowRecordingsFile | null = await getWorkflowById(
    workflowId
  );
  // TODO handle case where a workflow file is not found for the given workflowId
  if (workflowData === null) {
    throw new Error(
      `TODO handle case where a workflow file is not found for the given workflowId. workflowId:${workflowId}`
    );
  }

  const operationsWorkflowData = (workflowData as WorkflowRecordingsFileVersion1)
    .operationRecordings[opId];
  // TODO handle case where the workflow file does not have a recording for the given opId
  if (!operationsWorkflowData) {
    throw new Error(
      `TODO handle case where the workflow file does not have a recording for the given opId. opId:${opId}`
    );
  }

  return operationsWorkflowData.opRecordingId;
};

type GetRootTypeRecordingIdFromOpRecording = (params: {
  rootTypeId: string;
  opId: string;
  opRecordingId: string;
}) => Promise<string | null>;

export const getRootTypeRecordingIdFromOpRecording: GetRootTypeRecordingIdFromOpRecording = async ({
  rootTypeId,
  opId,
  opRecordingId,
}) => {
  // get opId.json
  const opFile: OperationRecordingsFileVersion1 | null = await getOperationFile(
    opId
  );

  if (opFile === null) {
    // TODO handle case where operation file does not exist for given operation id
    throw new Error(
      `TODO handle case where operation file does not exist for given operation id. operationId:${opId}`
    );
  }
  // return the recordingId associated with the given typeId
  const opRecording = opFile.recordings[opRecordingId];

  if (opRecording === undefined) {
    // TODO handle case where operation file does not have a recording for the given operation recording id
    throw new Error(
      `TODO handle case where operation file does not have a recording for the given operation recording id. opRecordingId:${opRecordingId}`
    );
  }

  const rootTypeRecordingEntry = opRecording.rootTypeRecordings[rootTypeId];

  if (rootTypeRecordingEntry === undefined) {
    // TODO handle case where the given operation recording does not have an entry for the given root type id
    throw new Error(
      `TODO handle case where the given operation recording does not have an entry for the given root type id.
        opRecordingId:${opRecordingId}
        rootTypeId:${rootTypeId}`
    );
  }

  return rootTypeRecordingEntry.recordingId || null;
};

export interface FetchRootTypeRecordingParams {
  opName: string;
  operationsData: OperationNameToIdMapping;
  rootTypeName: string;
  typeNameToIdMappingData: TypeNameToIdMapping;
  workflowId: string;
}

// TODO handle args
// TODO optimization: should only need to do this once for a given root type on a given operation
//  * wait for semaphore access (only allows a single access at any given time) https://www.npmjs.com/package/await-semaphore
//  * after acquiring semaphore, check cache for resolved value
//  * If exists, release semaphore and return value
//  * Else, do the work below, set cache, release semaphore, and return value
export const fetchRootTypeRecording = async ({
  opName,
  operationsData,
  rootTypeName,
  typeNameToIdMappingData,
  workflowId,
}: FetchRootTypeRecordingParams) => {
  const opId = getOpIdFromOpName({
    operationsData,
    opName,
  });

  const opRecordingId = await getOpRecordingIdFromWorkflow({
    opId,
    workflowId,
  });

  const rootTypeId = await getTypeIdFromTypeName({
    typeName: rootTypeName,
    typeNameToIdMappingData,
  });
  // TODO handle the case where the given typeName does not exist in the given schemaId's types.json
  if (rootTypeId === null) {
    throw new Error(`
      TODO handle the case where the given typeName does not exist in the given types.json.
      rootTypeName:${rootTypeName}
    `);
  }

  const rootTypeRecordingId = await getRootTypeRecordingIdFromOpRecording({
    rootTypeId,
    opId,
    opRecordingId,
  });

  if (rootTypeRecordingId === null) {
    // TODO handle case where typeRecordingId does not exist for the given rootTypeId, opId, opRecordingId
    throw new Error(
      `TODO handle case where typeRecordingId does not exist for the given rootTypeId, opId, opRecordingId.
        rootTypeId:${rootTypeId}
        opId:${opId}
        opRecordingId:${opRecordingId}`
    );
  }

  return await getTypeRecording({
    typeId: rootTypeId,
    recordingId: rootTypeRecordingId,
  });
};
