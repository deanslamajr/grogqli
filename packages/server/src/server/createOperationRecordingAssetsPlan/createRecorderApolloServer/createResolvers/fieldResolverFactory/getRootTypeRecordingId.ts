import { IntrospectionQuery } from 'graphql';

import { RootTypeRecordingsIds } from '../..';

type GetGenericRootTypeName = (params: {
  schema: IntrospectionQuery;
  rootTypeName: string;
}) => 'query' | 'mutation' | 'subscription';

const getGenericRootTypeName: GetGenericRootTypeName = ({
  schema,
  rootTypeName,
}) => {
  const { queryType, mutationType, subscriptionType } = schema.__schema;
  let genericRootTypeName;

  if (rootTypeName === queryType.name) {
    genericRootTypeName = 'query';
  } else if (rootTypeName === mutationType?.name) {
    genericRootTypeName = 'mutation';
  } else if (rootTypeName === subscriptionType?.name) {
    genericRootTypeName = 'subscription';
  } else {
    throw new Error(
      `updateRecordingsPlanForRootTypeField was invoked for a root type:(${rootTypeName}) not found in the given schema.`
    );
  }
  return genericRootTypeName;
};

type GetRootTypeRecordingId = (params: {
  rootTypeName: string;
  rootTypeRecordingsIds: RootTypeRecordingsIds;
  schema: IntrospectionQuery;
}) => string;

export const getRootTypeRecordingId: GetRootTypeRecordingId = ({
  rootTypeName,
  rootTypeRecordingsIds,
  schema,
}) => {
  const genericRootTypeName = getGenericRootTypeName({
    schema,
    rootTypeName,
  });
  return rootTypeRecordingsIds[genericRootTypeName];
};
