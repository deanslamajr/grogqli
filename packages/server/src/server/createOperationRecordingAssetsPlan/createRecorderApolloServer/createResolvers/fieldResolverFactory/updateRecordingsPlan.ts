import shortid from 'shortid';

import { OperationRecordingPlan } from '../..';
import { ReturnTypeInfo } from '.';

interface EncodedFieldResolverValue {
  value: any;
  parentTypeRecordingId: string;
}

export type UpdateRecordingsPlan = <T extends any>(params: {
  opName: string;
  parentTypeName: string;
  parentTypeRecordingId: string;
  recordingsPlan: OperationRecordingPlan;
  fieldName: string;
  fieldTypeInfo: ReturnTypeInfo;
  fieldValue: T;
}) => T | EncodedFieldResolverValue | EncodedFieldResolverValue[];

export const updateRecordingsPlan: UpdateRecordingsPlan = ({
  opName,
  parentTypeName,
  parentTypeRecordingId,
  recordingsPlan,
  fieldName,
  fieldTypeInfo,
  fieldValue,
}) => {
  if (recordingsPlan.name === undefined) {
    recordingsPlan.name = opName;
  }
  // if an entry does not yet exist for the given parentTypeRecordingId,
  // create a new one
  if (recordingsPlan.typeRecordings[parentTypeRecordingId] === undefined) {
    recordingsPlan.typeRecordings[parentTypeRecordingId] = {
      typeName: parentTypeName,
      value: {},
    };
  }

  let parentTypeRecordingFieldValue;
  let fieldResolverValue;

  // handle the different return type cases
  // 1. the field return type is a scalar or list of scalars (or the value is null/undefined)
  //  e.g. we don't want to create a type recording of this data, the parent type can own the data
  //  * return the fieldValue as is
  if (
    fieldTypeInfo.isScalar ||
    fieldValue === null ||
    fieldValue === undefined
  ) {
    parentTypeRecordingFieldValue = fieldValue;
    fieldResolverValue = fieldValue;
  }
  // 2. the field return type is a non-scalars or list of non-scalars
  // 2a. if list of scalars, return [
  //    {value: fieldValue[0], parentTypeRecordingId: recordingId1},
  //    {value: fieldValue[1], parentTypeRecordingId: recordingId2},
  //    etc
  //  ]
  else if (!fieldTypeInfo.isScalar && fieldTypeInfo.isList) {
    if (!Array.isArray(fieldValue)) {
      throw new Error(
        'Return type is a LIST type but the given non-null, defined field value is not an array type.'
      );
    }

    parentTypeRecordingFieldValue = [];
    fieldResolverValue = [];

    fieldValue.forEach((returnListItem) => {
      const recordingId = shortid.generate();
      parentTypeRecordingFieldValue.push(recordingId);

      const formattedReturnListItem: EncodedFieldResolverValue = {
        value: returnListItem,
        parentTypeRecordingId: recordingId,
      };
      fieldResolverValue.push(formattedReturnListItem);
    });
  }
  // 2b. if a single non-scalar (e.g. non-list), return {value: fieldValue, parentTypeRecordingId}
  else {
    const recordingId = shortid.generate();
    parentTypeRecordingFieldValue = recordingId;

    fieldResolverValue = {
      value: fieldValue,
      parentTypeRecordingId: recordingId,
    } as EncodedFieldResolverValue;
  }

  recordingsPlan.typeRecordings[parentTypeRecordingId].value[
    fieldName
  ] = parentTypeRecordingFieldValue;

  return fieldResolverValue;
};
