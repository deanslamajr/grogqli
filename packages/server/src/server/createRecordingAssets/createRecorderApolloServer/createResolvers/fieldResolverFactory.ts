import { GraphQLFieldResolver, IntrospectionQuery } from 'graphql';
import shortid from 'shortid';

import { Context, RecordingsPlan } from '..';

interface IsTopLevelFieldParams {
  schema: IntrospectionQuery;
  parentTypeName: string;
}

export interface ResolveValueFactoryParams {
  schema: IntrospectionQuery;
  parentTypeName: string;
  fieldName: string;
}

const isTopLevelField = ({
  schema,
  parentTypeName,
}: IsTopLevelFieldParams): boolean => {
  const { queryType, mutationType, subscriptionType } = schema.__schema;
  return (
    parentTypeName === queryType.name ||
    parentTypeName === mutationType?.name ||
    parentTypeName === subscriptionType?.name
  );
};

interface UpdateRecordingPlanParams {
  fieldName: string;
  fieldTypeName: string;
  fieldTypeRecordingId: string;
  fieldTypeRecording: any;
  recordingsPlan: RecordingsPlan;
  parentTypeName: string;
}
const updateRecordingPlan = ({
  fieldName,
  fieldTypeName,
  fieldTypeRecordingId,
  fieldTypeRecording,
  recordingsPlan,
  parentTypeName,
}: UpdateRecordingPlanParams): void => {
  // update the parent type's reference to this field
  // TODO check if is an array and if so, push the value into the array
  recordingsPlan.typeRecordings[parentTypeName].value[
    fieldName
  ] = fieldTypeRecordingId;
  // add this fields type recording
  recordingsPlan.typeRecordings[fieldTypeName] = {
    recordingId: fieldTypeRecordingId,
    value: fieldTypeRecording,
  };
};

export const fieldResolverFactory = ({
  schema,
  parentTypeName,
  fieldName,
}: ResolveValueFactoryParams): GraphQLFieldResolver<any, Context> => {
  return async (parent, args, context, info) => {
    const { recordingsPlan } = context.runTimeVariables.grogqli!;
    let responseRecordingSlice;

    // const opName = info.operation?.name?.value;
    // TODO handle unnamed query case
    // if (!opName) {
    //   throw new Error('TODO handle unnamed query case');
    // }

    const isRootField = isTopLevelField({ schema, parentTypeName });
    if (isRootField) {
      const parsedOpResponseRecording = JSON.parse(
        context.runTimeVariables.grogqli!.operationResponseRecording
      );
      responseRecordingSlice = parsedOpResponseRecording.data[fieldName];
    } else {
      responseRecordingSlice = parent[fieldName];
    }

    const typeNameOfField = responseRecordingSlice.__typename;
    delete responseRecordingSlice.__typename;
    const fieldTypeRecordingId: string = shortid.generate();

    updateRecordingPlan({
      fieldName,
      fieldTypeName: typeNameOfField,
      fieldTypeRecordingId,
      fieldTypeRecording: responseRecordingSlice,
      parentTypeName,
      recordingsPlan,
    });

    return responseRecordingSlice;
  };
};
