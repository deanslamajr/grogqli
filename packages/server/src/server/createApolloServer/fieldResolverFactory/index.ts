import { GraphQLFieldResolver, IntrospectionQuery } from 'graphql';

import { isRootType } from '../../isRootType';
import { TypeNameToIdMapping } from '../../files/type';
import { OperationNameToIdMapping } from '../../files/operation';
import { Context } from '..';

import { fetchRootTypeRecording } from './fetchRootTypeRecording';
import { fetchNestedTypeRecording } from './fetchNestedTypeRecording';

export interface ResolveValueFactoryParams {
  operationsData: OperationNameToIdMapping;
  typeNameToIdMappingData: TypeNameToIdMapping;
  schema: IntrospectionQuery;
  parentTypeName: string;
  fieldName: string;
}

export const fieldResolverFactory = ({
  operationsData,
  typeNameToIdMappingData,
  schema,
  parentTypeName,
  fieldName,
}: ResolveValueFactoryParams): GraphQLFieldResolver<any, Context> => {
  return async (parent, args, context, info) => {
    let fieldValueFromRecording;

    const isRootField = isRootType({ schema, typeName: parentTypeName });
    if (isRootField) {
      const opName = info.operation?.name?.value;
      // TODO handle unnamed query case
      if (!opName) {
        throw new Error('TODO handle unnamed query case');
      }

      const rootTypeRecording = await fetchRootTypeRecording({
        opName,
        operationsData,
        rootTypeName: parentTypeName,
        typeNameToIdMappingData,
        workflowId: context.runTimeVariables.grogqli!.workflowId,
      });

      // TODO (optionally) use args to destructure the recording associated with the given args
      fieldValueFromRecording = rootTypeRecording[fieldName];

      // TODO handle case where recording doesn't have a value for the given field
      if (fieldValueFromRecording === undefined) {
        throw new Error(
          `TODO handle case where recording doesn't have a value for the given field.
          parentTypeName:${parentTypeName} fieldName:${fieldName}`
        );
      }
    }

    // Nested field
    else {
      const nestedTypeRecording = await fetchNestedTypeRecording({
        recordingId: parent,
        typeName: parentTypeName,
        typeNameToIdMappingData,
      });

      // TODO handle case where recording doesn't have a value for the given field
      fieldValueFromRecording = nestedTypeRecording[fieldName];

      // TODO handle case when a fieldValue does not exist for the given typeRecording
      if (fieldValueFromRecording === undefined) {
        throw new Error(
          `TODO handle case when a fieldValue does not exist for the given typeRecording.
          parentTypeName:${parentTypeName} fieldName:${fieldName}`
        );
      }
    }

    return fieldValueFromRecording;
  };
};
