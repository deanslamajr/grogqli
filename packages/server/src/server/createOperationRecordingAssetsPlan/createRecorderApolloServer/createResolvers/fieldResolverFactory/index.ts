import {
  GraphQLFieldResolver,
  IntrospectionQuery,
  IntrospectionOutputTypeRef,
  FieldNode,
} from 'graphql';

import { isRootType } from '../../../../isRootType';
import { Context } from '../..';
import { getRootTypeRecordingId } from './getRootTypeRecordingId';
import { updateRecordingsPlan } from './updateRecordingsPlan';

export interface ResolveValueFactoryParams {
  returnType: IntrospectionOutputTypeRef;
  schema: IntrospectionQuery;
  parentTypeName: string;
  fieldName: string;
}

export interface ReturnTypeInfo {
  isList: boolean;
  isScalar: boolean;
}

type GetReturnTypeInfo = (
  returnType: IntrospectionOutputTypeRef
) => ReturnTypeInfo;

const getReturnTypeInfo: GetReturnTypeInfo = (returnType) => {
  let isList = false;
  let isScalar = false;

  function parseType(type: IntrospectionOutputTypeRef) {
    if (type.kind === 'SCALAR') {
      isScalar = true;
    } else if (type.kind === 'LIST') {
      isList = true;
      return parseType(type.ofType);
    } else if (type.kind === 'NON_NULL') {
      return parseType(type.ofType);
    }
  }

  parseType(returnType);

  return {
    isList,
    isScalar,
  };
};

type GetFieldNameOrAlias = (params: {
  fieldName: string;
  fieldNodes: ReadonlyArray<FieldNode>;
}) => string;

const getFieldNameOrAlias: GetFieldNameOrAlias = ({
  fieldName,
  fieldNodes,
}) => {
  const currentNode = fieldNodes.find(({ name }) => name.value === fieldName);

  if (currentNode !== undefined && currentNode.alias !== undefined) {
    return currentNode.alias.value;
  } else {
    return fieldName;
  }
};

export const fieldResolverFactory = ({
  returnType,
  schema,
  parentTypeName,
  fieldName,
}: ResolveValueFactoryParams): GraphQLFieldResolver<any, Context> => {
  return async (parent, args, context, info) => {
    const opName = info.operation?.name?.value;

    const fieldNameOrAlias = getFieldNameOrAlias({
      fieldName,
      fieldNodes: info.fieldNodes,
    });

    // TODO handle unnamed query case
    if (!opName) {
      throw new Error('TODO handle unnamed query case');
    }

    const {
      recordingsPlan,
      rootTypeRecordingsIds,
    } = context.runTimeVariables.grogqli!;
    let fieldResolverValue;

    const returnTypeInfo: ReturnTypeInfo = getReturnTypeInfo(returnType);

    const isRootField = isRootType({ schema, typeName: parentTypeName });
    if (isRootField) {
      // TODO handle case
      //  where context.runTimeVariables.grogqli!.parsedOpRecording.data
      //  === null || undefined
      const fieldValue = context.runTimeVariables.grogqli!.parsedOpRecording
        .data[fieldNameOrAlias];

      const rootTypeRecordingId = getRootTypeRecordingId({
        rootTypeName: parentTypeName,
        rootTypeRecordingsIds,
        schema,
      });

      fieldResolverValue = updateRecordingsPlan({
        opName,
        parentTypeName,
        recordingsPlan,
        fieldName,
        fieldTypeInfo: returnTypeInfo,
        fieldValue,
        parentTypeRecordingId: rootTypeRecordingId,
        isRootType: true,
      });
    } else {
      const fieldValue = parent.value[fieldNameOrAlias];
      const parentTypeRecordingId = parent.parentTypeRecordingId;

      fieldResolverValue = updateRecordingsPlan({
        opName,
        parentTypeName,
        recordingsPlan,
        fieldName,
        fieldTypeInfo: returnTypeInfo,
        fieldValue,
        parentTypeRecordingId,
      });
    }

    return fieldResolverValue;
  };
};
