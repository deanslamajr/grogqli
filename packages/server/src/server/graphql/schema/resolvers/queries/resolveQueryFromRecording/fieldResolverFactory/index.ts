import { GraphQLFieldResolver, IntrospectionQuery } from 'graphql';

import { OperationsData, TypeNameToIdMapping } from '../../../files';
import { Context } from '../createApolloServer';

import { fetchRootTypeRecording } from './fetchRootTypeRecording';
import { fetchNestedTypeRecording } from './fetchNestedTypeRecording';

interface IsTopLevelFieldParams {
  schema: IntrospectionQuery;
  parentTypeName: string;
}
const isTopLevelField = ({
  schema,
  parentTypeName,
}: IsTopLevelFieldParams): boolean => {
  const { queryType, mutationType, subscriptionType } = schema.__schema;
  return (
    parentTypeName === queryType.name ||
    parentTypeName === mutationType?.name ||
    parentTypeName === subscriptionType?.name
  );
};

export interface ResolveValueFactoryParams {
  operationsData: OperationsData;
  typeNameToIdMappingData: TypeNameToIdMapping;
  schema: IntrospectionQuery;
  parentTypeName: string;
  fieldName: string;
}
export const fieldResolverFactory = ({
  operationsData,
  typeNameToIdMappingData,
  schema,
  parentTypeName,
  fieldName,
}: ResolveValueFactoryParams): GraphQLFieldResolver<any, Context> => {
  return async (parent, args, context, info) => {
    let fieldValueFromRecording;

    const opName = info.operation?.name?.value;
    // TODO handle unnamed query case
    if (!opName) {
      throw new Error('TODO handle unnamed query case');
    }

    const isRootField = isTopLevelField({ schema, parentTypeName });
    if (isRootField) {
      const rootTypeRecording = await fetchRootTypeRecording({
        opName,
        operationsData,
        workflowId: context.runTimeVariables.grogqli!.workflowId,
      });

      // TODO (optionally) use args to destructure the recording associated with the given args
      fieldValueFromRecording = rootTypeRecording[fieldName];

      // TODO handle case where recording doesn't have a value for the given field
      if (fieldValueFromRecording === undefined) {
        throw new Error(
          `TODO handle case where recording doesn't have a value for the given field.
          parentTypeName:${parentTypeName} fieldName:${fieldName}`
        );
      }
    }

    // Nested field
    else {
      const nestedTypeRecording = await fetchNestedTypeRecording({
        recordingId: parent,
        typeName: parentTypeName,
        typeNameToIdMappingData,
      });

      // TODO handle case where recording doesn't have a value for the given field
      fieldValueFromRecording = nestedTypeRecording[fieldName];

      // TODO handle case when a fieldValue does not exist for the given typeRecording
      if (fieldValueFromRecording === undefined) {
        throw new Error(
          `TODO handle case when a fieldValue does not exist for the given typeRecording.
          parentTypeName:${parentTypeName} fieldName:${fieldName}`
        );
      }
    }

    return fieldValueFromRecording;
  };
};
