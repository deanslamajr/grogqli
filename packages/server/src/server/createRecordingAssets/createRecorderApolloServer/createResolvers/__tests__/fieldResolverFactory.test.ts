import { fieldResolverFactory } from '../fieldResolverFactory';
import { Context, RecordingsPlan } from '../../index';

import schemaFile from './schema.json';

jest.mock('shortid', () => ({
  generate: () => 'mockRecordingId',
}));

describe('fieldResolverFactory', () => {
  describe('the function created from the factory', () => {
    let fieldResolver: ReturnType<typeof fieldResolverFactory>;
    const schema = schemaFile.introspectionQuery;
    const info = {};
    const operationResponseRecording =
      '{"data":{"search":{"artists":{"nodes":[{"id":"QXJ0aXN0OmJjNzEwYmNmLTg4MTUtNDJjZi1iYWQyLTNmMWQxMjI0NmFlYg==","name":"Nickelback","__typename":"Artist"},{"id":"QXJ0aXN0OjFkNmI5M2JhLTg1MWUtNDQ3Yy1iOTg3LThkMzA4ZDZhYzg5ZA==","name":"A Motown Tribute to Nickelback","__typename":"Artist"}],"__typename":"ArtistConnection"},"__typename":"SearchQuery"}}}';

    describe('resolving the field of a root type', () => {
      const queryTypeName = schema.__schema.queryType.name;
      const fieldName = 'search';

      beforeEach(() => {
        fieldResolver = fieldResolverFactory({
          schema,
          parentTypeName: queryTypeName,
          fieldName,
        });
      });

      it('should update the recording plan', async () => {
        const recordingsPlan: RecordingsPlan = {
          typeRecordings: {
            [queryTypeName]: {
              recordingId: 'someRecordingId',
              value: {},
            },
          },
        };

        const context: Context = {
          schemaId: schemaFile.id,
          runTimeVariables: {
            grogqli: {
              operationResponseRecording,
              recordingsPlan,
            },
          },
        };

        await fieldResolver({}, {}, context, info);

        expect(recordingsPlan).toMatchSnapshot();
      });

      it('should return data for child resolvers', async () => {
        const recordingsPlan: RecordingsPlan = {
          typeRecordings: {
            [queryTypeName]: {
              recordingId: 'someRecordingId',
              value: {},
            },
          },
        };

        const context: Context = {
          schemaId: schemaFile.id,
          runTimeVariables: {
            grogqli: {
              operationResponseRecording,
              recordingsPlan,
            },
          },
        };

        const response = await fieldResolver({}, {}, context, info);

        expect(response).toMatchSnapshot();
      });
    });

    describe('resolving the field of a nested type', () => {
      const queryTypeName = schema.__schema.queryType.name;
      const fieldName = 'artists';
      const parentTypeName = 'SearchQuery';

      const dataFromParentResolver = {
        artists: {
          __typename: 'ArtistConnection',
          nodes: [
            {
              __typename: 'Artist',
              id:
                'QXJ0aXN0OmJjNzEwYmNmLTg4MTUtNDJjZi1iYWQyLTNmMWQxMjI0NmFlYg==',
              name: 'Nickelback',
            },
            {
              __typename: 'Artist',
              id:
                'QXJ0aXN0OjFkNmI5M2JhLTg1MWUtNDQ3Yy1iOTg3LThkMzA4ZDZhYzg5ZA==',
              name: 'A Motown Tribute to Nickelback',
            },
          ],
        },
      };

      beforeEach(() => {
        fieldResolver = fieldResolverFactory({
          schema,
          parentTypeName,
          fieldName,
        });
      });

      it('should update the recording plan', async () => {
        const recordingsPlan: RecordingsPlan = {
          typeRecordings: {
            [queryTypeName]: {
              recordingId: 'someRecordingId',
              value: {
                search: 'anotherRecordingId',
              },
            },
            [parentTypeName]: {
              recordingId: 'anotherRecordingId',
              value: {},
            },
          },
        };

        const context: Context = {
          schemaId: schemaFile.id,
          runTimeVariables: {
            grogqli: {
              operationResponseRecording,
              recordingsPlan,
            },
          },
        };

        await fieldResolver(dataFromParentResolver, {}, context, info);

        expect(recordingsPlan).toMatchSnapshot();
      });

      it('should return data for child resolvers', async () => {
        const recordingsPlan: RecordingsPlan = {
          typeRecordings: {
            [queryTypeName]: {
              recordingId: 'someRecordingId',
              value: {
                search: 'anotherRecordingId',
              },
            },
            [parentTypeName]: {
              recordingId: 'anotherRecordingId',
              value: {},
            },
          },
        };

        const context: Context = {
          schemaId: schemaFile.id,
          runTimeVariables: {
            grogqli: {
              operationResponseRecording,
              recordingsPlan,
            },
          },
        };

        const response = await fieldResolver(
          dataFromParentResolver,
          {},
          context,
          info
        );

        expect(response).toMatchSnapshot();
      });
    });
  });
});
