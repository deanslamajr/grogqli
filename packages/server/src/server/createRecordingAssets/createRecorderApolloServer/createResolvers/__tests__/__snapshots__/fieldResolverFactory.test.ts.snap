// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fieldResolverFactory the function created from the factory resolving the field of a nested type should return data for child resolvers 1`] = `
Object {
  "nodes": Array [
    Object {
      "__typename": "Artist",
      "id": "QXJ0aXN0OmJjNzEwYmNmLTg4MTUtNDJjZi1iYWQyLTNmMWQxMjI0NmFlYg==",
      "name": "Nickelback",
    },
    Object {
      "__typename": "Artist",
      "id": "QXJ0aXN0OjFkNmI5M2JhLTg1MWUtNDQ3Yy1iOTg3LThkMzA4ZDZhYzg5ZA==",
      "name": "A Motown Tribute to Nickelback",
    },
  ],
}
`;

exports[`fieldResolverFactory the function created from the factory resolving the field of a nested type should update the recording plan 1`] = `
Object {
  "typeRecordings": Object {
    "ArtistConnection": Object {
      "recordingId": "mockRecordingId",
      "value": Object {
        "nodes": Array [
          Object {
            "__typename": "Artist",
            "id": "QXJ0aXN0OmJjNzEwYmNmLTg4MTUtNDJjZi1iYWQyLTNmMWQxMjI0NmFlYg==",
            "name": "Nickelback",
          },
          Object {
            "__typename": "Artist",
            "id": "QXJ0aXN0OjFkNmI5M2JhLTg1MWUtNDQ3Yy1iOTg3LThkMzA4ZDZhYzg5ZA==",
            "name": "A Motown Tribute to Nickelback",
          },
        ],
      },
    },
    "Query": Object {
      "recordingId": "someRecordingId",
      "value": Object {
        "search": "anotherRecordingId",
      },
    },
    "SearchQuery": Object {
      "recordingId": "anotherRecordingId",
      "value": Object {
        "artists": "mockRecordingId",
      },
    },
  },
}
`;

exports[`fieldResolverFactory the function created from the factory resolving the field of a root type should return data for child resolvers 1`] = `
Object {
  "artists": Object {
    "__typename": "ArtistConnection",
    "nodes": Array [
      Object {
        "__typename": "Artist",
        "id": "QXJ0aXN0OmJjNzEwYmNmLTg4MTUtNDJjZi1iYWQyLTNmMWQxMjI0NmFlYg==",
        "name": "Nickelback",
      },
      Object {
        "__typename": "Artist",
        "id": "QXJ0aXN0OjFkNmI5M2JhLTg1MWUtNDQ3Yy1iOTg3LThkMzA4ZDZhYzg5ZA==",
        "name": "A Motown Tribute to Nickelback",
      },
    ],
  },
}
`;

exports[`fieldResolverFactory the function created from the factory resolving the field of a root type should update the recording plan 1`] = `
Object {
  "typeRecordings": Object {
    "Query": Object {
      "recordingId": "someRecordingId",
      "value": Object {
        "search": "mockRecordingId",
      },
    },
    "SearchQuery": Object {
      "recordingId": "mockRecordingId",
      "value": Object {
        "artists": Object {
          "__typename": "ArtistConnection",
          "nodes": Array [
            Object {
              "__typename": "Artist",
              "id": "QXJ0aXN0OmJjNzEwYmNmLTg4MTUtNDJjZi1iYWQyLTNmMWQxMjI0NmFlYg==",
              "name": "Nickelback",
            },
            Object {
              "__typename": "Artist",
              "id": "QXJ0aXN0OjFkNmI5M2JhLTg1MWUtNDQ3Yy1iOTg3LThkMzA4ZDZhYzg5ZA==",
              "name": "A Motown Tribute to Nickelback",
            },
          ],
        },
      },
    },
  },
}
`;
